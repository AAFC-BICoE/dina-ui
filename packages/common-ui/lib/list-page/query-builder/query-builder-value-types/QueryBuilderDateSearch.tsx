import React from "react";
import DatePicker from "react-datepicker";
import {
  includedTypeQuery,
  rangeQuery,
  termQuery,
  existsQuery
} from "../query-builder-elastic-search/QueryBuilderElasticSearchExport";
import { TransformToDSLProps } from "../../types";
import { DATE_REGEX_PARTIAL } from "common-ui/lib";

interface QueryBuilderDateSearchProps {
  /**
   * Current match type being used.
   */
  matchType?: string;

  /**
   * Retrieve the current value from the Query Builder.
   */
  value: string;

  /**
   * Pass the selected value to the Query Builder to store.
   */
  setValue: ((fieldPath: string) => void) | undefined;
}

export default function QueryBuilderDateSearch({
  matchType,
  value,
  setValue
}: QueryBuilderDateSearchProps) {
  return (
    <>
      {matchType !== "empty" && matchType !== "notEmpty" && (
        <DatePicker
          className="form-control"
          value={value}
          onChange={(newDate: Date, event) => {
            if (
              !event ||
              event?.type === "click" ||
              event?.type === "keydown"
            ) {
              setValue?.(newDate && newDate.toISOString().slice(0, 10));
            }
          }}
          onChangeRaw={(event) => {
            if (event?.type === "change") {
              let newText = event.target.value;
              const dashOccurrences = newText.split("-").length - 1;
              if (newText.length === 8 && dashOccurrences === 0) {
                newText =
                  newText.slice(0, 4) +
                  "-" +
                  newText.slice(4, 6) +
                  "-" +
                  newText.slice(6);
              }
              setValue?.(newText);
            }
          }}
          dateFormat="yyyy-MM-dd"
          placeholderText="YYYY-MM-DD"
          isClearable={true}
          showYearDropdown={true}
          todayButton="Today"
        />
      )}
    </>
  );
}

/**
 * Using the query row for a date search, generate the elastic search request to be made.
 */
export function transformDateSearchToDSL({
  operation,
  value,
  fieldInfo,
  fieldPath
}: TransformToDSLProps): any {
  if (!fieldInfo) {
    return {};
  }

  const { parentType, parentName } = fieldInfo;

  switch (operation) {
    // Contains / less than / greater than / less than or equal to / greater than or equal to.
    case "containsDate":
    case "greaterThan":
    case "greaterThanOrEqualTo":
    case "lessThan":
    case "lessThanOrEqualTo":
      return parentType
        ? {
            nested: {
              path: "included",
              query: {
                bool: {
                  must: [
                    rangeQuery(
                      fieldPath,
                      buildDateRangeObject(operation, value)
                    ),
                    includedTypeQuery(parentType)
                  ]
                }
              }
            }
          }
        : rangeQuery(fieldPath, buildDateRangeObject(operation, value));

    // Not equals match type.
    case "notEquals":
      return parentType
        ? {
            bool: {
              should: [
                // If the field does exist, then search for everything that does NOT match the term.
                {
                  nested: {
                    path: "included",
                    query: {
                      bool: {
                        must_not: termQuery(fieldPath, value, false),
                        must: includedTypeQuery(parentType)
                      }
                    }
                  }
                },

                // If it's included but the field doesn't exist, then it's not equal either.
                {
                  nested: {
                    path: "included",
                    query: {
                      bool: {
                        must_not: existsQuery(fieldPath),
                        must: includedTypeQuery(parentType)
                      }
                    }
                  }
                },

                // And if it's not included, then it's not equal either.
                {
                  bool: {
                    must_not: existsQuery(
                      "data.relationships." + parentName + ".data.id"
                    )
                  }
                }
              ]
            }
          }
        : {
            bool: {
              should: [
                {
                  bool: {
                    must_not: termQuery(fieldPath, value, false)
                  }
                },
                {
                  bool: {
                    must_not: existsQuery(fieldPath)
                  }
                }
              ]
            }
          };

    // Empty values only. (only if the value is not mandatory)
    case "empty":
      return parentType
        ? {
            bool: {
              should: [
                {
                  bool: {
                    must_not: {
                      nested: {
                        path: "included",
                        query: {
                          bool: {
                            must: [
                              existsQuery(fieldPath),
                              includedTypeQuery(parentType)
                            ]
                          }
                        }
                      }
                    }
                  }
                },
                {
                  bool: {
                    must_not: existsQuery(
                      "data.relationships." + parentName + ".data.id"
                    )
                  }
                }
              ]
            }
          }
        : {
            bool: {
              must_not: existsQuery(fieldPath)
            }
          };

    // Not empty values only. (only if the value is not mandatory)
    case "notEmpty":
      return parentType
        ? {
            nested: {
              path: "included",
              query: {
                bool: {
                  must: [existsQuery(fieldPath), includedTypeQuery(parentType)]
                }
              }
            }
          }
        : existsQuery(fieldPath);

    // Equals and default case
    default:
      return parentType
        ? {
            nested: {
              path: "included",
              query: {
                bool: {
                  must: [
                    termQuery(fieldPath, value, false),
                    includedTypeQuery(parentType)
                  ]
                }
              }
            }
          }
        : termQuery(fieldPath, value, false);
  }
}

/**
 * Depending on the numerical match type, the search query changes.
 *
 * Equal is ignored here since it should not be handled like this.
 *
 * Contains is a special case since it is not a date match, it treats it as a range of dates. For
 * example:
 *
 * "2022" will display all records that contain 2022 in the date field. So the following would be
 * matched:
 *    - 2022-01-01
 *    - 2022-12-02
 *    - 2022-05-19
 *    - 2022-07
 *    - 2022
 *
 * When using Equals to search for a date, the following would be matched for "2022":
 *    - 2022
 *
 * @param matchType the operator type (example: greaterThan ---> gt)
 * @param value The operator value to search against.
 * @returns numerical operator and value.
 */
function buildDateRangeObject(matchType, value) {
  switch (matchType) {
    case "containsDate":
      const YEAR_REGEX = /^\d{4}$/;
      const MONTH_REGEX = /^\d{4}-\d{2}$/;

      // Check if the value matches the year regex
      if (YEAR_REGEX.test(value)) {
        return {
          gte: `${value}||/y`,
          lte: `${value}||/y`,
          format: "yyyy"
        };
      }

      // Check if the value matches the month regex
      if (MONTH_REGEX.test(value)) {
        return {
          gte: `${value}||/M`,
          lte: `${value}||/M`,
          format: "yyyy-MM"
        };
      }

      // Otherwise just try to match the full date provided.
      return {
        gte: `${value}||/d`,
        lte: `${value}||/d`,
        format: "yyyy-MM-dd"
      };
    case "greaterThan":
      return { gt: value };
    case "greaterThanOrEqualTo":
      return { gte: value };
    case "lessThan":
      return { lt: value };
    case "lessThanOrEqualTo":
      return { lte: value };
    default:
      return value;
  }
}

/**
 * Validate the date string to ensure it's something elastic search can accept.
 *
 * Partial dates are supported here.
 * @param value date value
 * @param formatMessage error message translation locale
 * @return null if valid, string error if not valid.
 */
export function validateDate(value, formatMessage): string | null {
  if (DATE_REGEX_PARTIAL.test(value)) {
    return null;
  }
  return formatMessage({ id: "dateMustBeFormattedPartial" });
}
