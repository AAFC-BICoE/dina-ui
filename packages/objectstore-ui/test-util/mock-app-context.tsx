import {
  AccountContextI,
  AccountProvider,
  ApiClientContextI,
  AuthenticatedApiClientProvider,
  createContextValue,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { ObjectStoreIntlProvider } from "../intl/objectstore-intl";

interface MockAppContextProviderProps {
  apiContext?: Partial<ApiClientContextI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <AccountProvider
      value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
    >
      <AuthenticatedApiClientProvider
        apiContext={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContext)}
      >
        <ObjectStoreIntlProvider>
          <ModalProvider appElement={document.querySelector("body")}>
            {children}
          </ModalProvider>
        </ObjectStoreIntlProvider>
      </AuthenticatedApiClientProvider>
    </AccountProvider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}

const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = {
  authenticated: true,
  groups: ["/aafc", "cnc"],
  initialized: true,
  login: noop,
  logout: noop,
  token: "test-token",
  username: "test-user"
};

const DEFAULT_API_CONTEXT_VALUE = createContextValue({
  getTempIdGenerator: () => () => "00000000-0000-0000-0000-000000000000"
});
