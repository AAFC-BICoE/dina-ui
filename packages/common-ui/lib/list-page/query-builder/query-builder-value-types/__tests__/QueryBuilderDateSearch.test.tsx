import { mountWithAppContext } from "common-ui/lib/test-util/mock-app-context";
import QueryBuilderDateSearch, {
  transformDateSearchToDSL,
  validateDate
} from "../QueryBuilderDateSearch";

describe("QueryBuilderDateSearch", () => {
  describe("QueryBuilderDateSearch Component", () => {
    it("Display field if match type is equals", async () => {
      // This test will just ensure the layout does not change unexpectedly.
      // Any changes to the layout, the snapshots will need to be updated.
      const dateSearchEquals = mountWithAppContext(
        <QueryBuilderDateSearch
          matchType="equals"
          value="test"
          setValue={jest.fn}
        />
      );

      // Expect a snapshot with the date field being displayed.
      expect(
        dateSearchEquals.find(QueryBuilderDateSearch).debug()
      ).toMatchSnapshot(
        "Expect date field to be displayed since match type is equals"
      );

      const dateSearchEmpty = mountWithAppContext(
        <QueryBuilderDateSearch
          matchType="empty"
          value="test"
          setValue={jest.fn}
        />
      );

      // Expect a snapshot without the date field being displayed.
      expect(
        dateSearchEmpty.find(QueryBuilderDateSearch).debug()
      ).toMatchSnapshot(
        "Expect date field not to be displayed since the match type is not equals"
      );
    });
  });

  describe("transformDateSearchToDSL function", () => {
    describe("ContainsDate operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "containsDate",
            value: "1998",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "containsDate"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        // Contains the year only.
        expect(
          transformDateSearchToDSL({
            operation: "containsDate",
            value: "1998",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "containsDate"
          })
        ).toMatchSnapshot();

        // Contains the year and month only.
        expect(
          transformDateSearchToDSL({
            operation: "containsDate",
            value: "1998-05",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "containsDate"
          })
        ).toMatchSnapshot();

        // Contains the full date.
        expect(
          transformDateSearchToDSL({
            operation: "containsDate",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "containsDate"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Greater Than operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "greaterThan",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "greaterThan"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "greaterThan",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "greaterThan"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Greater Than Or Equal To operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "greaterThanOrEqualTo",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "greaterThanOrEqualTo"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "greaterThanOrEqualTo",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "greaterThanOrEqualTo"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Less Than operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "lessThan",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "lessThan"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "lessThan",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "lessThan"
          })
        ).toMatchSnapshot();
      });
    });

    describe("lessThanOrEqualTo operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "lessThanOrEqualTo",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "lessThanOrEqualTo"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "lessThanOrEqualTo",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "lessThanOrEqualTo"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Equals operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "equals",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "equals",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Partial Match operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "partialMatch",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "partialMatch"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "partialMatch",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "partialMatch"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Exact Match operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "exactMatch",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "exactMatch",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Not Equals operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "notEquals",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "notEquals"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "notEquals",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "notEquals"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Empty operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "empty",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "empty"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "empty",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "empty"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Not empty operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "notEmpty",
            value: "1998-05-19",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "notEmpty"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "notEmpty",
            value: "1998-05-19",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "notEmpty"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Equals operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "equals",
            value: "false",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "equals",
            value: "true",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.dateField",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Edge cases", () => {
      test("If no field value is provided, nothing should be generated.", async () => {
        expect(
          transformDateSearchToDSL({
            operation: "exactMatch",
            value: "false",
            fieldInfo: undefined,
            fieldPath: "includes.name",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });
    });
  });

  describe("validateDate function", () => {
    test("Correct date formats", async () => {
      // Null is expected, since it will return an error message if incorrect.
      expect(validateDate("1998", jest.fn)).toBeNull();
      expect(validateDate("1998-05", jest.fn)).toBeNull();
      expect(validateDate("1998-05-19", jest.fn)).toBeNull();
    });

    test("Invalid date formats", async () => {
      expect(validateDate("19-05-1998", jest.fn)).not.toBeNull();
      expect(validateDate("today", jest.fn)).not.toBeNull();
      expect(validateDate("December 19th 2022", jest.fn)).not.toBeNull();
    });
  });
});
