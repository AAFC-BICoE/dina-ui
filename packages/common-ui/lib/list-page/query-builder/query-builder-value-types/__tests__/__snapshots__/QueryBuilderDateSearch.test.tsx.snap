// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilderDateSearch QueryBuilderDateSearch Component Display field if match type is equals: Expect date field not to be displayed since the match type is not equals 1`] = `"<QueryBuilderDateSearch matchType=\\"empty\\" value=\\"test\\" setValue={[Function: bound fn]} />"`;

exports[`QueryBuilderDateSearch QueryBuilderDateSearch Component Display field if match type is equals: Expect date field to be displayed since match type is equals 1`] = `
"<QueryBuilderDateSearch matchType=\\"equals\\" value=\\"test\\" setValue={[Function: bound fn]}>
  <r className=\\"form-control\\" value=\\"test\\" onChange={[Function: onChange]} onChangeRaw={[Function: onChangeRaw]} dateFormat=\\"yyyy-MM-dd\\" placeholderText=\\"YYYY-MM-DD\\" isClearable={true} showYearDropdown={true} todayButton=\\"Today\\" allowSameDay={false} dateFormatCalendar=\\"LLLL yyyy\\" disabled={false} disabledKeyboardNavigation={false} dropdownMode=\\"scroll\\" onFocus={[Function: onFocus]} onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyDown]} onInputClick={[Function: onInputClick]} onSelect={[Function: onSelect]} onClickOutside={[Function: onClickOutside]} onMonthChange={[Function: onMonthChange]} onCalendarOpen={[Function: onCalendarOpen]} onCalendarClose={[Function: onCalendarClose]} preventOpenOnFocus={false} onYearChange={[Function: onYearChange]} onInputError={[Function: onInputError]} monthsShown={1} readOnly={false} withPortal={false} shouldCloseOnSelect={true} showTimeSelect={false} showTimeInput={false} showPreviousMonths={false} showMonthYearPicker={false} showFullMonthYearPicker={false} showTwoColumnMonthYearPicker={false} showFourColumnMonthYearPicker={false} showYearPicker={false} showQuarterYearPicker={false} strictParsing={false} timeIntervals={30} timeCaption=\\"Time\\" previousMonthAriaLabel=\\"Previous Month\\" previousMonthButtonLabel=\\"Previous Month\\" nextMonthAriaLabel=\\"Next Month\\" nextMonthButtonLabel=\\"Next Month\\" previousYearAriaLabel=\\"Previous Year\\" previousYearButtonLabel=\\"Previous Year\\" nextYearAriaLabel=\\"Next Year\\" nextYearButtonLabel=\\"Next Year\\" timeInputLabel=\\"Time\\" enableTabLoop={true} yearItemNumber={12} renderDayContents={[Function: renderDayContents]} focusSelectedMonth={false} showPopperArrow={true} excludeScrollbar={true} customTimeInput={{...}} calendarStartDay={[undefined]}>
    <r className={[undefined]} wrapperClassName={[undefined]} hidePopper={true} portalId={[undefined]} popperModifiers={{...}} targetComponent={{...}} popperContainer={[undefined]} popperComponent={{...}} popperPlacement=\\"bottom-start\\" popperProps={{...}} popperOnKeyDown={[Function (anonymous)]} enableTabLoop={true}>
      <Manager className=\\"react-datepicker-manager\\">
        <Reference>
          <div className=\\"react-datepicker-wrapper\\">
            <div className=\\"react-datepicker__input-container\\">
              <input type=\\"text\\" value=\\"test\\" onBlur={[Function (anonymous)]} onChange={[Function (anonymous)]} onClick={[Function (anonymous)]} onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} id={[undefined]} name={[undefined]} autoFocus={[undefined]} placeholder=\\"YYYY-MM-DD\\" disabled={false} autoComplete={[undefined]} className=\\"form-control\\" title={[undefined]} readOnly={false} required={[undefined]} tabIndex={[undefined]} aria-describedby={[undefined]} aria-invalid={[undefined]} aria-labelledby={[undefined]} aria-required={[undefined]} />
            </div>
          </div>
        </Reference>
      </Manager>
    </r>
  </r>
</QueryBuilderDateSearch>"
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function ContainsDate operation Normal field 1`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "format": "yyyy",
      "gte": "1998||/y",
      "lte": "1998||/y",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function ContainsDate operation Normal field 2`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "format": "yyyy-MM",
      "gte": "1998-05||/M",
      "lte": "1998-05||/M",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function ContainsDate operation Normal field 3`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "format": "yyyy-MM-dd",
      "gte": "1998-05-19||/d",
      "lte": "1998-05-19||/d",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function ContainsDate operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "range": Object {
              "includes.name": Object {
                "format": "yyyy",
                "gte": "1998||/y",
                "lte": "1998||/y",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Edge cases If no field value is provided, nothing should be generated. 1`] = `Object {}`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Empty operation Normal field 1`] = `
Object {
  "bool": Object {
    "must_not": Object {
      "exists": Object {
        "field": "data.attributes.dateField",
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Empty operation With relationship as field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "must_not": Object {
            "nested": Object {
              "path": "included",
              "query": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "exists": Object {
                        "field": "includes.name",
                      },
                    },
                    Object {
                      "term": Object {
                        "included.type": "collection",
                      },
                    },
                  ],
                },
              },
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.relationships.collection.data.id",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Equals operation Normal field 1`] = `
Object {
  "term": Object {
    "data.attributes.dateField": "1998-05-19",
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Equals operation Normal field 2`] = `
Object {
  "term": Object {
    "data.attributes.dateField": "true",
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Equals operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "includes.name": "1998-05-19",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Equals operation With relationship as field 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "includes.name": "false",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Exact Match operation Normal field 1`] = `
Object {
  "term": Object {
    "data.attributes.dateField": "1998-05-19",
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Exact Match operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "includes.name": "1998-05-19",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Greater Than Or Equal To operation Normal field 1`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "gte": "1998-05-19",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Greater Than Or Equal To operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "range": Object {
              "includes.name": Object {
                "gte": "1998-05-19",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Greater Than operation Normal field 1`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "gt": "1998-05-19",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Greater Than operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "range": Object {
              "includes.name": Object {
                "gt": "1998-05-19",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Less Than operation Normal field 1`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "lt": "1998-05-19",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Less Than operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "range": Object {
              "includes.name": Object {
                "lt": "1998-05-19",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Not Equals operation Normal field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "must_not": Object {
            "term": Object {
              "data.attributes.dateField": "1998-05-19",
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.attributes.dateField",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Not Equals operation With relationship as field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "nested": Object {
          "path": "included",
          "query": Object {
            "bool": Object {
              "must": Object {
                "term": Object {
                  "included.type": "collection",
                },
              },
              "must_not": Object {
                "term": Object {
                  "includes.name": "1998-05-19",
                },
              },
            },
          },
        },
      },
      Object {
        "nested": Object {
          "path": "included",
          "query": Object {
            "bool": Object {
              "must": Object {
                "term": Object {
                  "included.type": "collection",
                },
              },
              "must_not": Object {
                "exists": Object {
                  "field": "includes.name",
                },
              },
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.relationships.collection.data.id",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Not empty operation Normal field 1`] = `
Object {
  "exists": Object {
    "field": "data.attributes.dateField",
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Not empty operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "exists": Object {
              "field": "includes.name",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Partial Match operation Normal field 1`] = `
Object {
  "term": Object {
    "data.attributes.dateField": "1998-05-19",
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function Partial Match operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "includes.name": "1998-05-19",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function lessThanOrEqualTo operation Normal field 1`] = `
Object {
  "range": Object {
    "data.attributes.dateField": Object {
      "lte": "1998-05-19",
    },
  },
}
`;

exports[`QueryBuilderDateSearch transformDateSearchToDSL function lessThanOrEqualTo operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "range": Object {
              "includes.name": Object {
                "lte": "1998-05-19",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;
